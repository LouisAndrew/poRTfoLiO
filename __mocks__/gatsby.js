const React = require('react')
const gatsby = jest.requireActual('gatsby')

const mockSocials = [
    { link: 'http://localhost', social_key: 'github' },
    { link: 'http://google.com', social_key: 'dribbble' },
]

const mockButtonText = 'Click me'
const mockContactContent = 'Content'
const mockContactHeadline = 'Headline'
const mockEmail = 'john@doe.com'

module.exports = {
    ...gatsby,
    graphql: jest.fn(),
    Link: jest.fn().mockImplementation(
        // these props are invalid for an `a` tag
        ({
            activeClassName,
            activeStyle,
            getProps,
            innerRef,
            ref,
            replace,
            to,
            ...rest
        }) =>
            React.createElement('a', {
                ...rest,
                href: to,
            })
    ),
    StaticQuery: jest.fn(),
    // can I mock return value here?
    useStaticQuery: jest.fn(() => ({
        socialData: {
            frontmatter: {
                social: mockSocials,
            },
        },
        emailData: {
            frontmatter: {
                contact: {
                    button_text: mockButtonText,
                    contact_content: mockContactContent,
                    contact_headline: mockContactHeadline,
                    email: mockEmail,
                },
            },
        },

        // forced to write the hero-page-query mock here.
        heroData: {
            frontmatter: {
                hero: {
                    hero_content: 'Hello, World!', // hard coded from hero.test.tsx
                    hero_headline: 'I am headline', // Also hard coded from hero.test.tsx
                },
            },
        },
        imgS: {
            frontmatter: {
                hero: {
                    hero_imgs: [
                        {
                            childImageSharp: {
                                fixed: {},
                            },
                        },
                    ],
                },
            },
        },
        imgM: {
            frontmatter: {
                hero: {
                    hero_imgs: [
                        {
                            childImageSharp: {
                                fixed: {},
                            },
                        },
                    ],
                },
            },
        },
        imgL: {
            frontmatter: {
                hero: {
                    hero_imgs: [
                        {
                            childImageSharp: {
                                fixed: {},
                            },
                        },
                    ],
                },
            },
        },
        imgXL: {
            frontmatter: {
                hero: {
                    hero_imgs: [
                        {
                            childImageSharp: {
                                fixed: {},
                            },
                        },
                    ],
                },
            },
        },
    })),
}
