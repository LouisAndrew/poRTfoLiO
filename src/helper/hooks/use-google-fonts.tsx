import { useEffect } from 'react'

/**
 * Custom hook to embed google font script / api to html head
 */
const useGoogleFonts = () => {
    const LINK_URL =
        'https://fonts.googleapis.com/css2?family=Karla:wght@400;700;800&family=Nunito:wght@300;400;700;800&display=swap'

    const LINK_PRECONNECT_URL = 'https://fonts.gstatic.com'

    /**
     * Function to create link elements
     * @returns array of link element
     */
    const createLinks = () => {
        const fontLink = document.createElement('link')
        const preconnectLink = document.createElement('link')

        fontLink.href = LINK_URL
        fontLink.rel = 'stylesheet'

        preconnectLink.href = LINK_PRECONNECT_URL
        preconnectLink.rel = 'preconnect'

        return [fontLink, preconnectLink]
    }

    useEffect(() => {
        const links = createLinks()

        // Append links to document's head tag
        links.forEach(link => {
            document.head.appendChild(link)
        })

        return () => {
            // cleanup, remove links from document's head tag
            links.forEach(link => {
                document.head.removeChild(link)
            })
        }
    }, [])
}

export { useGoogleFonts }
