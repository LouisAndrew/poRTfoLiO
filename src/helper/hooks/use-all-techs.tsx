import { useEffect, useState } from 'react'
import { useStaticQuery, graphql } from 'gatsby'
import { FixedObject } from 'gatsby-image'

import { Tech } from 'helper/schema/Tech'

const useAllTechs = () => {
    const data = useStaticQuery(graphql`
        query {
            techData: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_name
                        }
                    }
                }
            }
            imgXS: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_logo {
                                childImageSharp {
                                    fixed(width: 40, height: 40, quality: 100) {
                                        ...GatsbyImageSharpFixed
                                    }
                                }
                            }
                        }
                    }
                }
            }
            imgS: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_logo {
                                childImageSharp {
                                    fixed(width: 60, height: 60, quality: 100) {
                                        ...GatsbyImageSharpFixed
                                    }
                                }
                            }
                        }
                    }
                }
            }
            imgM: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_logo {
                                childImageSharp {
                                    fixed(
                                        width: 100
                                        height: 100
                                        quality: 100
                                    ) {
                                        ...GatsbyImageSharpFixed
                                    }
                                }
                            }
                        }
                    }
                }
            }
            imgL: allMarkdownRemark(
                filter: { frontmatter: { template_key: { eq: "tech" } } }
            ) {
                edges {
                    node {
                        frontmatter {
                            tech_logo {
                                childImageSharp {
                                    fixed(
                                        width: 120
                                        height: 120
                                        quality: 100
                                    ) {
                                        ...GatsbyImageSharpFixed
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    `)

    const [techs, setTechs] = useState<Tech[]>([])

    useEffect(() => {
        const { techData, imgS, imgM, imgL, imgXS } = data as any

        console.log({ techData, imgS, imgM, imgL, imgXS })

        const temp: Tech[] = techData.edges.map((edge: any, i: number) => {
            const techName = edge.node.frontmatter.tech_name
            const xs =
                imgXS.edges[i].node.frontmatter.tech_logo.childImageSharp.fixed
            const s =
                imgS.edges[i].node.frontmatter.tech_logo.childImageSharp.fixed
            const m =
                imgM.edges[i].node.frontmatter.tech_logo.childImageSharp.fixed
            const l =
                imgL.edges[i].node.frontmatter.tech_logo.childImageSharp.fixed
            const xl: FixedObject[] = []

            return {
                techName,
                techLogo: {
                    xs,
                    s,
                    m,
                    l,
                    xl,
                },
            }
        })

        setTechs(temp)
    }, [])

    /**
     * Get a single tech by filtering its name
     * @param name Name of the technology
     * @returns the queried tech, and null if the tech is not available..
     */
    const getTechByName = (name: string) => {
        const techArr = techs.filter(
            tech => tech.techName.toLowerCase() === name.toLowerCase()
        )

        if (techArr.length === 0) {
            return null
        } else {
            return techArr[0]
        }
    }

    return { getTechByName, techs }
}

export { useAllTechs }
