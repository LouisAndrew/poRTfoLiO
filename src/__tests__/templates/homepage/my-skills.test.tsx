import React from 'react'
import ReactDOM from 'react-dom'
// import renderer from 'react-test-renderer'

// import { render, cleanup } from '@testing-library/react'
import { cleanup } from '@testing-library/react'
import '@testing-library/jest-dom'

import MySkills from 'templates/homepage/my-skills'

import * as Gatsby from 'gatsby'
import * as CustomHook from 'helper/hooks/use-all-techs'

const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery')
const useAllTechs = jest.spyOn(CustomHook, 'useAllTechs')

const mockSkills = ['React', 'Typescript']
const mockTechs = [
    {
        techName: mockSkills[0],
        techLogo: {
            s: [],
            xs: [],
            m: [],
            l: [],
            xl: [],
        },
    },
    {
        techName: mockSkills[1],
        techLogo: {
            s: [],
            xs: [],
            m: [],
            l: [],
            xl: [],
        },
    },
]

// TODO: how to mock useStaticQuery and custom hook at the same time?
describe('MySkills', () => {
    const Element = <MySkills />

    afterEach(cleanup)
    beforeEach(() => {
        useAllTechs.mockImplementationOnce(() => ({
            techs: mockTechs,
            getTechByName: name => {
                if (name === mockSkills[0]) {
                    return mockTechs[0]
                } else {
                    return mockTechs[1]
                }
            },
        }))

        useStaticQuery.mockImplementationOnce(() => ({
            markdownRemark: {
                frontmatter: {
                    skills: mockSkills,
                },
            },
            techData: {
                edges: [
                    {
                        node: {
                            frontmatter: {
                                tech_name: mockSkills,
                            },
                        },
                    },
                ],
            },
        }))
    })

    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDOM.render(Element, div)
    })

    // Failure in mocking custom hook and staticquery in the document at the same time..

    // it('Should render all fetched skill cards correctly', () => {
    //     const { queryByTestId } = render(Element)

    //     mockSkills.forEach(name => {
    //         const testId = `homepage-skill-${name}`

    //         const el = queryByTestId(testId)
    //         expect(el).toBeInTheDocument()
    //     })
    // })

    // it('matches snapshot', () => {
    //     const run = true

    //     expect(run).toBeTruthy()

    //     if (run) {
    //         const tree = renderer.create(Element).toJSON()
    //         expect(tree).toMatchSnapshot()
    //     }
    // })
})
