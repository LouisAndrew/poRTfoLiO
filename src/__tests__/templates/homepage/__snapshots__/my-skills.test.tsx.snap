// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MySkills matches snapshot 1`] = `
<div
  className="css-4cffwv"
  id="skills"
>
  <div
    className="css-16vu25q"
  >
    <h2
      className="css-l7njxq"
    >
      Some Things I'm Familiar With
    </h2>
    <div
      className="css-1o29yoi"
    >
      <div
        className="skillcard css-zg82jx"
        data-testid="wrapper"
      >
        <div
          className="css-1vyba89"
        >
          <div
            className=" gatsby-image-wrapper"
            style={
              Object {
                "display": "inline-block",
                "height": undefined,
                "overflow": "hidden",
                "position": "relative",
                "width": undefined,
              }
            }
          >
            <picture>
              <source
                media="(max-width: 640px)"
              />
              <source
                media="(max-width: 832px) and (orientation: landscape)"
              />
              <source
                media="(max-width: 1040px)"
              />
              <source
                media="(min-width: 1952px)"
              />
              <source />
              <img
                alt=""
                loading="lazy"
                onLoad={[Function]}
                style={
                  Object {
                    "height": "auto",
                    "left": 0,
                    "objectFit": "cover",
                    "objectPosition": "center",
                    "opacity": 0,
                    "position": "absolute",
                    "top": 0,
                    "transition": "opacity 500ms",
                    "width": "100%",
                  }
                }
              />
            </picture>
            <noscript
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<picture><source media=\\"(max-width: 640px)\\" srcset=\\"undefined\\" /><source media=\\"(max-width: 832px) and (orientation: landscape)\\" srcset=\\"undefined\\" /><source media=\\"(max-width: 1040px)\\" srcset=\\"undefined\\" /><source media=\\"(min-width: 1952px)\\" srcset=\\"undefined\\" /><source srcset=\\"undefined\\" /><img loading=\\"lazy\\" src=\\"\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
                }
              }
            />
          </div>
        </div>
        <div
          className="css-xr35o1"
        >
          React
        </div>
      </div>
    </div>
  </div>
</div>
`;
